Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/DbCommandBuilders/Specifications/SqlTableValuedParameterComparedColumnSpecificationTest.cs

```diff
@@ -0,0 +1,116 @@
+  private object[] _guidCollection;
+  private Mock<ISqlDialect> _sqlDialectMock;
+  private SqlCommand _sqlCommand;
+  private SqlParameter _sqlParameter;
+
+  [SetUp]
+  public void SetUp ()
+  {
+    var storageTypeInfo = StorageTypeInformationObjectMother.CreateUniqueIdentifierStorageTypeInformation();
+    _columnDefinition = new ColumnDefinition("Column", storageTypeInfo, false);
+    _guidCollection =
+    [
+        new Guid("C1BF6D17-6D4B-468D-928B-5E0D7A42F275"),
+        new Guid("22858375-AEB1-4F4F-9138-16F02DA3BAA0"),
+        new Guid("D47EF6FE-F64D-4EAE-B70F-2E06814F7674")
+    ];
+
+    _sqlCommand = new SqlCommand();
+    _sqlParameter = _sqlCommand.CreateParameter();
+
+    _sqlDialectMock = new Mock<ISqlDialect>(MockBehavior.Strict);
+  }
+
+  [Test]
+  public void AddParameters_WithGuidColumn_AddsTableValuedParameter ()
+  {
+    _sqlDialectMock.Setup(_ => _.GetParameterName(_columnDefinition.Name)).Returns("@param");
+    _sqlDialectMock.Setup(_ => _.CreateDataParameter(_sqlCommand, _columnDefinition.StorageTypeInfo, "@param", It.IsAny<object>())).Returns(_sqlParameter);
+
+    var specification = new SqlTableValuedParameterComparedColumnSpecification(_columnDefinition, _guidCollection);
+    specification.AddParameters(_sqlCommand, _sqlDialectMock.Object);
+
+    Assert.That(_sqlCommand.Parameters.Count, Is.EqualTo(1));
+    Assert.That(_sqlCommand.Parameters[0], Is.SameAs(_sqlParameter));
+
+    Assert.That(_sqlParameter.SqlDbType, Is.EqualTo(SqlDbType.Structured));
+    Assert.That(_sqlParameter.TypeName, Is.EqualTo("TVP_Guid"));
+    Assert.That(_sqlParameter.Value, Is.InstanceOf<SqlTableValuedParameterValue>());
+
+    var tvpValue = _sqlParameter.Value.As<SqlTableValuedParameterValue>();
+
+    var sqlMetadata = tvpValue.ColumnMetaData.Single();
+    Assert.That(sqlMetadata.Name, Is.EqualTo("Value"));
+    Assert.That(sqlMetadata.SqlDbType, Is.EqualTo(SqlDbType.UniqueIdentifier));
+
+    var items = tvpValue.Select(record => record.GetGuid(0));
+    Assert.That(items, Is.EquivalentTo(_guidCollection));
+  }
+
+  [Test]
+  public void AddParameters_WithEmptyCollection_AddsDBNull ()
+  {
+    _sqlDialectMock.Setup(_ => _.GetParameterName(_columnDefinition.Name)).Returns("@param");
+    _sqlDialectMock.Setup(_ => _.CreateDataParameter(_sqlCommand, _columnDefinition.StorageTypeInfo, "@param", It.IsAny<object>())).Returns(_sqlParameter);
+
+    var specification = new SqlTableValuedParameterComparedColumnSpecification(_columnDefinition, Array.Empty<object>());
+    specification.AddParameters(_sqlCommand, _sqlDialectMock.Object);
+
+    Assert.That(_sqlCommand.Parameters.Count, Is.EqualTo(1));
+    Assert.That(_sqlCommand.Parameters[0], Is.SameAs(_sqlParameter));
+
+    Assert.That(_sqlParameter.SqlDbType, Is.EqualTo(SqlDbType.Structured));
+    Assert.That(_sqlParameter.TypeName, Is.EqualTo("TVP_Guid"));
+    Assert.That(_sqlParameter.Value, Is.EqualTo(DBNull.Value));
+  }
+
+  [Test]
+  public void AppendComparisons ()
```

Comment:
i think we need a second test for emptycollection

Reply:
I don't think so, because at this point, the collection of values is hidden behind a parameter, and the SQL command text is the same for empty and non-empty collections.
Empty collections must be handled differently when creating the parameter, because ADO.NET cannot deal with an empty `SqlDataRecord[]`, but that limitation only affects the `SqlParameter.Value`.

Reply:
mhm... i'm not sure where it would fit, but some kind of itnegration test (or larger-scope unit test) would definitely be warrented to show everything works well together. It felt like here would be a good spot for thsi and the test would be cheap. you have a better / more meaningful place? 

And yes, there would be some redundancy, but the empty/null case is so special that it feels like it should be covered explicitly.

oh, and another reason why "here" would be good: we already test empty with the parameters, so testing with comparision, too, would be symetrical. I'm speaking from the experience of having looked for some documentation of expected behavior and tests are extremly helpful on this. So, cheap redundancy gets trumped by "acts as good docs".